---
title: "Plot sites of positive selection on alignment"
output: html_notebook
---

Set up libraries
```{r}
library(ape)
library(phangorn)
```

Plot aa positions.
```{r}
aln=read.dna('~/Documents/dawe_ab10_kindr/positive_selection/allsp_kindrlogs_entirealn.fasta', format='fasta', as.matrix=T)

aa=trans(aln)
image(aa, cex=0.5)


```
Read in the BUSTED results and plot them on the alignment. 
```{r, echo=TRUE, warning=FALSE}
upstream=read.csv('~/Documents/dawe_ab10_kindr/positive_selection/upstream_busted.csv')
downstream=read.csv('~/Documents/dawe_ab10_kindr/positive_selection/downstream_busted.csv')
### where these fit in aa positions of entire alignment
uRange=c(1:3,10,13:15,21:24,27:30,32:47,49,60:66,71:73,75:76,81:82,107,113,rep(NA,41), rep(NA, 113),201:935)  ### equal to 113 on upstream only alignment

dRange=c(89:935)  # range of entire aln that is same as downstream (where we will plot downstream on entire)

```

At site 114, we switch to the conserved region that is idential between both alignments.

```{r}
plot(upstream$site_index, upstream$er_constrained)
points(downstream$site_index, downstream$er_constrained, col='red')

```


```{r}
#library(extrafont)
#loadfonts()
#pdf('~/Downloads/positive_selection_along_gene.pdf', family='Arial', 8,6)
par(mfrow=c(2,1), mar=c(2,9,1,1))
image(aa, cex=0.5, legend=F)

 psr <- par("usr")
 xx <- psr[2]/2
 yy <- psr[4] * (0.5 + 0.5/par("plt")[4])
 legend(xx, yy, legend = c('Hydrophobic', 'Small', 'Hydrophilic'), pch = 22, pt.bg = c('red', 'yellow', 'blue'),
    pt.cex = 2, bty = "n", xjust = 0.5, yjust = 0.5,
    horiz = TRUE, xpd = TRUE)
#points(rep(17, 200),1:200)
#par(mar=c(2,7.95,1,1))
plot(1, 1, col=0, ylim=c(0,6), xlim=c(0,ncol(aa)), ylab='ln(evidence ratio for pos \nsel site under constrained model)', xaxs='i')
points(dRange, rev(downstream$er_constrained), col='red', pch=19, cex=0.5)
#points(c(1:48,201:935), rev(upstream$er_constrained)[uRange[!is.na(uRange)]], col='blue', pch=19)
points(1:48, rev(upstream$er_constrained)[c(1:3,10,13:15,21:24,27:30,32:47,49,60:66,71:73,75:76,81:82,107,113)], col='blue', pch=19, cex=0.5 )
points(202:935, rev(upstream$er_constrained)[114:847], col='blue', pch=19, cex=0.5)

## orthologous positions
points(dRange[14], rev(downstream$er_constrained)[14], cex=0.7, col='purple', pch=19)
points(6, rev(upstream$er_constrained)[14], cex=0.7, col='purple', pch=19)
abline(h=0.5, lty=2)

legend('topright', c('upstream exon1 duplication', 'downstream exon1 duplication'), pch=19, col = c('blue', 'red'), cex=0.8)
#dev.off()
```

For reference the categories each AA is assigned to are:

- Hydrophobic = .AA_raw[c("V", "I", "L", "F", "W", "Y", "M")],
- Small = .AA_raw[c("P", "G", "A", "C")],
- Hydrophilic = .AA_raw[c("S", "T", "H", "N", "Q", "D", "E", "K", "R")],
- Missing = .AA_raw[c('X')]
